rfcs/2022-12-09-541-reload-api.md

```diff
@@ -0,0 +1,277 @@
+
+Another option would be to branch earlier in `Application::run`, make options
+related to the control server exclusive with normal options like `--config`,
+and have more of a separate mode of operation for this functionality. This
+would remove some concerns about potential strange interactions between
+multiple paths for reloading (e.g. watching a file while running a control
+server while running an HTTP provider) and potentially require less refactoring
+work than what's proposed in this RFC.
+
+This approach wasn't chosen because the divergence for "normal" Vector
+operation would itself be a path for potentially confusing differences in
+behavior. It also would not move us closer to an improved overall design for
+this portion of Vector, and would create another path that we would later
+likely want to unify.
+
+### gRPC instead of JSON
+
+Instead of a REST-ish JSON API, we could use gRPC. We already use the
+technology in other areas of Vector (native source and sink), so it wouldn't
+have the same downside as GraphQL, while providing some of the benefits of an
+IDL.
+
+This approach was not chosen for a couple of reasons:
+- Protobuf adds complexity (i.e. code generation) to any client implementation.
+- The ability to interact with the API from simple tools like cURL is lost.
+- The only current endpoint accepts a Vector config, for which we already have
+  a JSON Schema and we would not want to translate to Protobuf.
+- Many of the benefits of gRPC are around high performance, load balancing,
+  health checking, etc, none of which is relevant to a small local control API.
+
+## Outstanding Questions
+
+- What is the expected flow for setting global options at startup (especially
+  those that can't be changed via reload)? Use normal file loading with only
+  global options set?
+- Should the API accept entire configs, or should we attempt to limit to just
+  sources/transforms/sinks?
+
+## Plan Of Attack
+
+- [ ] Extract controller struct to consolidate things that change with reloads
+- [ ] Write simple Unix socket HTTP server task with `POST /config` route that builds/validates config
+- [ ] Wrap and share controller struct with server task so it can actually do reloads
+- [ ] Refactor topology reload method to return errors rather than log directly
+
+## Future Improvements
+
+- Write an RFC for differentiating different levels of configuration (e.g.
+  global vs component, reloadable vs not) and modeling them better within
+  Vector. Then use this to have a clearer distinction between what should be
+  present in a global config file vs what's loadable via the API, make things
+  appropriately exclusive and reduce the number of possible weird interactions
+  from running multiple config loading mechanisms concurrently.
+- Migrate features like `top` and `tap` to use the new API and reduce
+  maintenance cost of old + new API.
+- Improve the internal capabilities/modularity of reloads and expose that via
```

Comment:
The current plan for validating a config before reporting on initial startup involves starting `vector` and seeing that the process has started successfully (implementation details TBD). Is there any way we could handle this via the API in the future? 

Not sure how feasible this is as I imagine it would require being able to start Vector without a config, idling until a config is provided.

Reply:
Could we just start Vector with an "empty" config (in this case, one that only includes the enterprise internal metrics pipeline) and then load in the user-provided config via the API? In that case it wouldn't need to be any different from normal reloads.

Once we have a more isolated "API mode" for running Vector, I do think it'd make sense to let it start up and continue running with a totally empty config.

Reply:
Ah, yea great point. I hadn't thought about giving it the internal enterprise config to start. That sounds good.